File: qsentinel/ledger/merkle.py
"""RFC 6962-style Merkle trees (the same construction as Certificate Transparency), with SHA3-256.

Leaves and inner nodes are domain-separated (0x00 / 0x01 prefix), so a leaf can never be
passed off as an inner node (second-preimage attack). An inclusion proof is O(log n) hashes:
anyone holding only the anchored root can check that a verdict is in the batch, without
downloading the ledger. That makes a "light client" auditor possible.
"""

from __future__ import annotations

import hashlib

EMPTY_ROOT = hashlib.sha3_256(b"").hexdigest()


def _leaf(data: bytes) -> bytes:
    return hashlib.sha3_256(b"\x00" + data).digest()


def _node(left: bytes, right: bytes) -> bytes:
    return hashlib.sha3_256(b"\x01" + left + right).digest()


def _split(n: int) -> int:
    """Largest power of two strictly less than n."""
    k = 1
    while k << 1 < n:
        k <<= 1
    return k


def _mth(leaves: list[bytes]) -> bytes:
    if len(leaves) == 1:
        return _leaf(leaves[0])
    k = _split(len(leaves))
    return _node(_mth(leaves[:k]), _mth(leaves[k:]))


def merkle_root(leaves: list[bytes]) -> str:
    return _mth(leaves).hex() if leaves else EMPTY_ROOT


def inclusion_proof(index: int, leaves: list[bytes]) -> list[tuple[str, str]]:
    """Audit path for leaves[index]: list of (side, sibling_hex); side 'L' = sibling on the left."""
    if not 0 <= index < len(leaves):
        raise IndexError(index)

    def path(m: int, d: list[bytes]) -> list[tuple[str, str]]:
        if len(d) == 1:
            return []
        k = _split(len(d))
        if m < k:
            return path(m, d[:k]) + [("R", _mth(d[k:]).hex())]
        return path(m - k, d[k:]) + [("L", _mth(d[:k]).hex())]

    return path(index, leaves)


def verify_inclusion(leaf: bytes, proof: list[tuple[str, str]] | list[list[str]], root: str) -> bool:
    h = _leaf(leaf)
    for side, sibling in proof:
        s = bytes.fromhex(sibling)
        h = _node(s, h) if side == "L" else _node(h, s)
    return h.hex() == root

File: web/src/lib/merkle.ts
// In-browser Merkle inclusion check, same construction as qsentinel/ledger/merkle.py (RFC 6962
// style, SHA3-256, 0x00 leaf / 0x01 node prefixes). Lets anyone verify a proof without trusting us.
import { sha3_256 } from "@noble/hashes/sha3.js";
import { bytesToHex, hexToBytes } from "@noble/hashes/utils.js";

function concat(...parts: Uint8Array[]) {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let o = 0;
  for (const p of parts) {
    out.set(p, o);
    o += p.length;
  }
  return out;
}

const leaf = (data: Uint8Array) => sha3_256(concat(new Uint8Array([0]), data));
const node = (l: Uint8Array, r: Uint8Array) => sha3_256(concat(new Uint8Array([1]), l, r));

export interface ProofStep { side: string; sibling: string; result: string }

/** Recomputes the root from the leaf and audit path; returns every intermediate hash. */
export function walkProof(leafHex: string, proof: [string, string][]): ProofStep[] {
  let h = leaf(hexToBytes(leafHex));
  const steps: ProofStep[] = [{ side: "leaf", sibling: "", result: bytesToHex(h) }];
  for (const [side, sib] of proof) {
    const s = hexToBytes(sib);
    h = side === "L" ? node(s, h) : node(h, s);
    steps.push({ side, sibling: sib, result: bytesToHex(h) });
  }
  return steps;
}

export function verifyProof(leafHex: string, proof: [string, string][], root: string): boolean {
  const steps = walkProof(leafHex, proof);
  return steps[steps.length - 1].result === root;
}
